include/ctype.h

```diff
@@ -65,13 +94,10 @@
 #ifdef __cplusplus
 static inline int isspace(int c)
 {
-  return c == ' ' || c == '\t' || c == '\n' || c == '\r' ||
-         c == '\f' || c == '\v';
+  return __ctype_lookup(c) & _S;
 }
 #else
-#  define isspace(c) \
-    ((c) == ' '  || (c) == '\t' || (c) == '\n' || (c) == '\r' || \
-     (c) == '\f' || (c) == '\v')
+#  define isspace(c) (__ctype_lookup(c) & _S)
```

Comment:
For the PROTECTED build, this will draw in two copies of the lookup table.  That doubles the FLASH overhead.

I think that we should condition this on defined(CONFIG_BUILD_FLAT) || !defined(__KERNEL__).  In the PROTECTED kernel build there should be very few ctyple reference and I think that they should be handled in a way that does not cause such a large overhead.

Reply:
> For the PROTECTED build, this will draw in two copies of the lookup table. That doubles the FLASH overhead.
> 
Yes, but all libc funcitons called by both userspace and kernel duplicate twice, the lookup table don't generate the huge difference here.

> I think that we should condition this on defined(CONFIG_BUILD_FLAT) || !defined(**KERNEL**). In the PROTECTED kernel build there should be very few ctyple reference and I think that they should be handled in a way that does not cause such a large overhead.

But this make the same code behaviour different in user space and kernel space. And does we(or user) really care about 256 ROM bytes addition? because the size explosion is much bigger than 256 bytes after we switch from flat mode to kernel/procted mode.

Reply:
> But this make the same code behaviour different in user space and kernel space. And does we(or user) really care about 256 ROM bytes addition? because the size explosion is much bigger than 256 bytes after we switch from flat mode to kernel/protected mode.

I am not sure if that is a good argument or not.  Systems become bloated and too large over time due to relatively small code size increased like this.  The only way to prevent code bloat is to be vigilant to keep the code growth down.

There should be neglible use of ctype functions within the OS.  Very little string processing is done by the OS itself.  So it seems very wasteful to force such a large step in code side for functionality that is not used much in the OS.

Reply:
But the standard compliant is the most INVIOLABLES rule. To achieve this, we even don't allow the user to turn off signal and file system as I remember correctly. The correction is always the first priority, if the size is more important than speed, I would suggest to change all macro to normal function. otherwise we can keep the macro but add the table to avoid double evaluation. The mix just make the thing worse.

Reply:
> There should be neglible use of ctype functions within the OS. Very little string processing is done by the OS itself. So it seems very wasteful to force such a large step in code side for functionality that is not used much in the OS.

If we ignore libs/ and tools/, there is almost no use of ctype in the OS.  Perhaps we should have a kctype.h header file for these:

   ./boards/mips/pic32mx/sure-pic32mx/src/pic32mx_lcd1602.c:#include <ctype.h>
   ./boards/renesas/m16c/skp16c26/src/m16c_lcd.c:#include <ctype.h>
   ./drivers/audio/tone.c:#include <ctype.h>
   ./drivers/syslog/syslog_rpmsg.c:#include <ctype.h>
   ./fs/dirent/fs_opendir.c:#include <ctype.h>
   ./fs/fat/fs_fat32dirent.c:#include <ctype.h>

There might be other sneak paths through libc that include ctype.h, but generally I would say that is a pretty high price to pay to such little value to the OS.  

The libc usage of ctype is not in functions normally used by the OS either.

Reply:
> But the standard compliant is the most INVIOLABLES rule.

We do not have to follow any external standards within the OS.  POSIX and OpenGroup.org ONLY define the interface between applications and the OS.  Within the OS we may do whatever we need to do to best support deeply embedded solutions.  So the INVIOLABLES.txt statement is irrelevant in the particular discussion.

Reply:
Linux works the same way.  It does not use the standard libc header files.  Rather, it is has its own set of header files that are in parallel with the libc header files with alternative definitions for use within the kernel.  It is a very standard thing to do within operating systems.

Reply:
> > There should be neglible use of ctype functions within the OS. Very little string processing is done by the OS itself. So it seems very wasteful to force such a large step in code side for functionality that is not used much in the OS.
> 
> If we ignore libs/ and tools/, there is almost no use of ctype in the OS. Perhaps we should have a kctype.h header file for these. That avoids the mix you dislike:
> 
> ./boards/mips/pic32mx/sure-pic32mx/src/pic32mx_lcd1602.c:#include <ctype.h>
> ./boards/renesas/m16c/skp16c26/src/m16c_lcd.c:#include <ctype.h>
> ./drivers/audio/tone.c:#include <ctype.h>
> ./drivers/syslog/syslog_rpmsg.c:#include <ctype.h>
> ./fs/dirent/fs_opendir.c:#include <ctype.h>
> ./fs/fat/fs_fat32dirent.c:#include <ctype.h>
> 
> There might be other sneak paths through libc that include ctype.h, but generally I would say that is a pretty high price to pay to such little value to the OS.
> 
> The libc usage of ctype is not in functions normally used by the OS either.

If so, how about change all macro to normal function? There are at least two perfect solution to resolve this issue:
1.Change all macro to inline function
2.Use the compiler extension:
   https://stackoverflow.com/questions/1238016/are-compound-statements-blocks-surrounded-by-parens-expressions-in-ansi-c
But we can't use them due to the old compiler. In any case, the correction is always the first priority. If the result is wrong, the size or speed don't have any meaning.

Reply:
> Linux works the same way. It does not use the standard libc header files. Rather, it is has its own set of header files that are in parallel with the libc header files with alternative definitions for use within the kernel. It is a very standard thing to do within operating systems.

We can use the special header file if you like, but the double evaluation must to address. no people want to spend one/two day to debug this type of error.

Reply:
> So the INVIOLABLES.txt statement is irrelevant in the particular discussion.

I meant that the first section of the INVIOLABLES.txt related to conforming to the portable POSIX application interface are not relative to this discussion.  Certainly, the section on modularity and the internal architecture of the OS always relevant to a discussion of internal OS changes.